### DISCUSSION TITLE
Source generated SearchValues?

### DISCUSSION THREAD

**Discussion by pawchen** [dotnet/runtime#95179] [General] [answered]
I don't know, it might be stupid. If Regex can benefit from compile-time generated codes, and it's starting to use SearchValues, why not put those analysis of SearchValues.Create in compile time?

**Comment by stephentoub** [comment, marked as answer]
> it might be stupid. If Regex can benefit from compile-time generated codes, and it's starting to use SearchValues, why not put those analysis of SearchValues.Create in compile time?

The regex source generator will emit use of SearchValues.

If you instead mean why does it emit a call to Create rather than emitting the logic behind Create, there's a lot of complicated logic there and it would require making public all the implementation details that back it, which we'd like to be able to (and plan to) evolve over time, and which includes a ton of vectorization-related logic we don't want to bake into every assembly. 

It's not stupid. Your suggestion is actually where we started, and then we evolved it to SearchValues due to all the complications.

**Reply by pawchen** [reply]
> If you instead mean why does it emit a call to Create rather than emitting the logic behind Create

Yes I mean the resulting tables/trees/lookups, or whatever decided as the best approach after calling Create. I think you get my point.

> we'd like to be able to (and plan to) evolve over time, and which includes a ton of vectorization-related logic we don't want to bake into every assembly

Ah I see, so users can benefit from improvements in the algorithm without recompiling when a newer version of the runtime is available. That makes sense! Thanks for the answer!

**Reply by gfoidl** [reply]
@pawchen when you're interested you can read through https://github.com/dotnet/runtime/issues/68328

**Reply by pawchen** [reply]
@gfoidl you actually implemented that, amazing! This is exactly the discussion I missed, thank you for sharing!
